### PR TITLE
MONGOCRYPT-702 FLE2 range floating point correctness issue when using precision mode

### PR DESCRIPTION
**Pull request by shreyaskalyan**

### CODE REVIEW

**Review by erwee — CHANGES_REQUESTED** [review]

**Review by shreyaskalyan — COMMENTED** [review]
RFAL. I updated the errors for decimal128. PTAL at subtraction closely in Decimal128. I think it's correct but maybe worth making it more resilient.

**Review by erwee — CHANGES_REQUESTED** [review]

File: src/mc-range-encoding.c [hunk 1]
```diff
@@ -345,42 +442,143 @@ static mlib_int128 dec128_to_int128(mc_dec128 dec) {
...
+        CLIENT_ERR("Invalid lower bound for Decimal128 precision. Fractional digits must be less than "
+                   "the specified precision value. min: %s, precision: %" PRIu32,
+                   mc_dec128_to_string(min).str,
+                   precision);
+        return false;
+    }
+
+    if (mc_dec128_greater(mc_dec128_abs(scaled_max), SIGNED_INT_128_MAX_DECIMAL)) {
+        CLIENT_ERR("Invalid upper bound for Decimal128 precision. Absolute scaled value must be less than "
+                   "%s. max: %s",
```
[the remark this answers is not in the document]
Comment (shreyaskalyan) [reply src/mc-range-encoding.c:535]: Done.

File: src/mc-range-encoding.c [hunk 2]
```diff
@@ -345,42 +442,143 @@ static mlib_int128 dec128_to_int128(mc_dec128 dec) {
...
+                   "%s. min: %s",
+                   mc_dec128_to_string(SIGNED_INT_128_MAX_DECIMAL).str,
+                   mc_dec128_to_string(min).str);
+        return false;
+    }
+
+    mc_dec128 t_1 = mc_dec128_sub(scaled_max, scaled_min);
+    mc_dec128 t_4 = mc_dec128_sub(UNSIGNED_INT_128_MAX_DECIMAL, t_1);
+    mc_dec128 t_5 =
+        mc_dec128_sub(mc_dec128_round_integral_ex(mc_dec128_log10(t_4), MC_DEC128_ROUND_UPWARD, NULL), MC_DEC128(1));
```
[the remark this answers is not in the document]
Comment (shreyaskalyan) [reply src/mc-range-encoding.c:552]: Done.

File: src/mc-range-encoding.c [hunk 3]
```diff
@@ -345,42 +442,143 @@ static mlib_int128 dec128_to_int128(mc_dec128 dec) {
...
+
+    // Because we have guaranteed earlier that max is greater than min, we can
+    // subtract these values and guarantee that taking their unsigned
+    // representation will yield the actual range result.
+    mlib_int128 range128 = mlib_int128_sub(i_1, i_2);
+
+    if (precision > 255) {
+        CLIENT_ERR("Invalid value for precision. Must be less than 255. precision: %" PRIu32, precision);
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (shreyaskalyan) [reply src/mc-range-encoding.c:570]: `mlib_int128_pow10` takes a uint8_t, so we have to ensure that precision can fit into there. I have changed it to uint8_t max.

File: src/mc-range-encoding.c [hunk 4]
```diff
@@ -345,42 +442,146 @@ static mlib_int128 dec128_to_int128(mc_dec128 dec) {
     return ret;
 }
 
-bool mc_canUsePrecisionModeDecimal(mc_dec128 min, mc_dec128 max, uint32_t precision, uint32_t *maxBitsOut) {
+#define SIGNED_INT_128_MAX_DECIMAL mc_dec128_from_string("170141183460469231731687303715884105727")
+#define UNSIGNED_INT_128_MAX_DECIMAL mc_dec128_from_string("340282366920938463463374607431768211455")
```
Comment (erwee) [inline src/mc-range-encoding.c:446]: Add comments for context. Also (nit), I prefer we use "(U)INT128_MAX_AS_DECIMAL128" for the naming.
```suggestion
// (2^127 - 1) = the maximum signed 128-bit integer value, as a decimal128
#define INT128_MAX_AS_DECIMAL128 mc_dec128_from_string("170141183460469231731687303715884105727")
// (2^128 - 1) = the max unsigned 128-bit integer value, as a decimal128
#define UINT128_MAX_AS_DECIMAL128 mc_dec128_from_string("340282366920938463463374607431768211455")
```

File: src/mc-range-encoding.c [hunk 5]
```diff
@@ -345,42 +442,146 @@ static mlib_int128 dec128_to_int128(mc_dec128 dec) {
...
+                   "or equal to %s. min: %s",
+                   mc_dec128_to_string(SIGNED_INT_128_MAX_DECIMAL).str,
+                   mc_dec128_to_string(min).str);
+        return false;
     }
 
-    return use_precision_mode;
+    mc_dec128 t_1 = mc_dec128_sub(scaled_max, scaled_min);
+    mc_dec128 t_4 = mc_dec128_sub(UNSIGNED_INT_128_MAX_DECIMAL, t_1);
+    mc_dec128 t_5 = mc_dec128_sub(mc_dec128_round_integral_ex(mc_dec128_log10(t_4), MC_DEC128_ROUND_TOWARD_ZERO, NULL),
```
Comment (erwee) [inline src/mc-range-encoding.c:551]: nit add comment
```suggestion
    // t_5 = floor(log10(t_4)) - 1;
    mc_dec128 t_5 = mc_dec128_sub(mc_dec128_round_integral_ex(mc_dec128_log10(t_4), MC_DEC128_ROUND_TOWARD_ZERO, NULL),
```

**Review by erwee — APPROVED** [review]

File: src/mc-range-encoding.c (hunk 4 above)
Comment (shreyaskalyan) [reply src/mc-range-encoding.c]: Done.

File: src/mc-range-encoding.c (hunk 5 above)
Comment (shreyaskalyan) [reply src/mc-range-encoding.c:555]: Done.

**Pull request merged**